### PR TITLE
Hypothesis implementation and other features

### PR DESCRIPTION
**Pull request by juloberno**
- Hypothesis implementation and behavior spaces
- benchmark runner: merging of intermediate checkpoints, allow to split result, histories and configs into individual files with partial loading of these files
- ACC behavior model implementation integrated into BaseIDM
- Distribution Class with parameterizable distributions

### CODE REVIEW

File: bark/benchmark/tests/py_benchmark_analyzer_tests.py [hunk 1]
```diff
@@ -4,19 +4,29 @@
 # This work is licensed under the terms of the MIT license.
 # For a copy, see <https://opensource.org/licenses/MIT>.
 
+try:
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_analyzer_tests.py:7]: why is this done with a try? can we not always import this?

File: bark/benchmark/tests/py_benchmark_process_tests.py [hunk 2]
```diff
@@ -22,12 +28,14 @@
...
-os.chdir("../benchmark_database/")
+
 
 class DatabaseRunnerTests(unittest.TestCase):
     def test_database_run_and_analyze(self):
-        dbs = DatabaseSerializer(test_scenarios=4, test_world_steps=2, num_serialize_scenarios=2)
+        dbs = DatabaseSerializer(test_scenarios=2, test_world_steps=3, num_serialize_scenarios=2)
+        # to find database files
+        cwd = os.getcwd()
+        os.chdir("../benchmark_database/")
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_process_tests.py:38]: why is the os.chdir necessary now? It does not seem to provide the same usability as before

File: bark/benchmark/tests/py_benchmark_process_tests.py [hunk 3]
```diff
@@ -42,12 +50,15 @@ def test_database_run_and_analyze(self):
...
+                                           log_eval_avg_every=2)
 
         result = benchmark_runner.run(maintain_history=True)
 
-        result.dump(os.path.join("./benchmark_results.pickle"))
-        result_loaded = BenchmarkResult.load(os.path.join("./benchmark_results.pickle"))
+        result.dump(os.path.join("./benchmark_results"), dump_configs=True, \
+                         dump_histories=True, max_mb_per_file=1)
+        result_loaded = BenchmarkResult.load(os.path.join("./benchmark_results"))
+        result_loaded.load_histories()
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_process_tests.py:60]: do we not also want to use camel-case in general for Python?

Comment (patrickhart) [reply bark/benchmark/tests/py_benchmark_process_tests.py:60]: I think it is rather confusing when switching between c++ and python

File: bark/benchmark/tests/py_benchmark_result_tests.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+          self.config_idx == other.config_idx
+def random_benchmark_conf_data(num_confs, conf_size):
+    confs = [ TestConfig(i, conf_size)for i in range(0, num_confs)]
+    return confs
+
+class DatabaseRunnerTests(unittest.TestCase):
+    def test_dump_and_load_results(self):
+        result_data = random_result_data(size = 10)
+        br = BenchmarkResult(result_dict=result_data)
+        br.dump("./results")
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_result_tests.py:44]: do we not need to delete this dump after the tests have finished? Consider adding a cleanup function?

File: bark/benchmark/tests/py_benchmark_runner_tests.py [hunk 5]
```diff
@@ -1,13 +1,17 @@
-# Copyright (c) 2020 Julian Bernhard, Klemens Esterle, Patrick Hart and
-# Tobias Kessler
+# Copyright (c) 2019 fortiss GmbH
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_runner_tests.py:1]: copyright have the wrong year and format.

File: bark/benchmark/tests/py_benchmark_result_tests.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) 2019 fortiss GmbH
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_result_tests.py:1]: please change

File: bark/benchmark/benchmark_result.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
+# Copyright (c) 2019 fortiss GmbH
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:1]: change copyright

File: bark/benchmark/BUILD [hunk 8]
```diff
@@ -5,10 +5,17 @@ py_library(
     imports = ['../../../python'],
     deps = [
         "//bark/runtime:runtime",
+        ":benchmark_result"
```
Comment (patrickhart) [inline bark/benchmark/BUILD:8]: I think paths in respect to the root folder are better as this allows moving code more easily.

File: bark/benchmark/benchmark_result.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+import logging
+import re
+import zipfile
+import math
+
+logging.basicConfig()
+logging.getLogger().setLevel(logging.INFO)
+
+
+class BehaviorConfig:
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:18]: maybe add some documentation to especially the scenario generation.
It is not as easy to grasp.

File: bark/benchmark/benchmark_result.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+import os
+import pickle
+import pandas as pd
+import logging
+import re
+import zipfile
+import math
+
+logging.basicConfig()
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:14]: why do we for example not use GLogInit here?

File: bark/benchmark/benchmark_result.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        for behavior_name, behavior in behavior_dict.items():
+            config = BehaviorConfig(behavior_name, behavior)
+            behavior_configs.append(config)
+        return behavior_configs
+
+# contains information for a single benchmark run
+class BenchmarkConfig:
+    def __init__(self, config_idx, behavior_config,
+                 scenario, scenario_idx, scenario_set_name):
+        self.config_idx = config_idx
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:40]: i plead for member variables to be like self._config_idx. Makes the separation more evident.

File: bark/benchmark/benchmark_result.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        self.behavior_config = behavior_config
+        self.scenario = scenario
+        self.scenario_idx = scenario_idx
+        self.scenario_set_name = scenario_set_name
+
+    def get_info_string_list(self):
+        info_strings = ["ConfigIdx: {}".format(self.config_idx),
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:47]: consider using f-strings in general

File: bark/benchmark/benchmark_result.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+                "scen_idx": self.scenario_idx,
+                **self.behavior_config.as_dict()}
+
+    def get_evaluation_groups(self):
+      return ["scen_set", *list(self.behavior_config.as_dict().keys())]
+
+
+# result of benchmark run
+class BenchmarkResult:
+    def __init__(self, result_dict = None, benchmark_configs = None, histories = None, data_frame = None, file_name = None):
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:65]: format

File: bark/benchmark/benchmark_result.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+                              histories=None)
+            self.extend(new_result)
+        return processed_files
+
+    @staticmethod
+    def _find_files_to_load(total_file_list, filetype, config_idx_list):
+        files_to_load = []
+        configs_not_found = set(config_idx_list)
+        for file in total_file_list:
+            match = re.search("config_idx_(?P<from>[0-9]+)_to_(?P<to>[0-9]+).{}".format(filetype), file)
```
Comment (patrickhart) [inline bark/benchmark/benchmark_result.py:182]: without documentation hard to understand

File: bark/benchmark/benchmark_runner_mp.py [hunk 15]
```diff
@@ -35,66 +39,98 @@ def deserialize_scenario(sc):
 # actor class running on a single core
 @ray.remote
 class _BenchmarkRunnerActor(BenchmarkRunner):
-    def __init__(self, evaluators, terminal_when, benchmark_configs, logger_name, log_eval_avg_every):
-          super().__init__(evaluators=evaluators, 
-                            terminal_when=terminal_when,
-                            benchmark_configs=benchmark_configs,
-                            logger_name=logger_name,
-                            log_eval_avg_every=log_eval_avg_every)
+    def __init__(self, evaluators, terminal_when, benchmark_configs, logger_name, log_eval_avg_every, checkpoint_dir, actor_id, glog_init_settings=None):
```
Comment (patrickhart) [inline bark/benchmark/benchmark_runner_mp.py:42]: format

File: bark/benchmark/tests/py_benchmark_runner_tests.py [hunk 16]
```diff
@@ -23,84 +27,217 @@
...
+        benchmark_runner = BenchmarkRunner(benchmark_database=db,
+                                           evaluators=evaluators,
+                                           terminal_when=terminal_when,
+                                           behaviors=behaviors_tested,
+                                           log_eval_avg_every=5)
+
+        result = benchmark_runner.run()
+        df = result.get_data_frame()
+        print(df)
+        self.assertEqual(len(df.index), 2*2*2) # 2 Behaviors * 2 Serialize Scenarios * 1 scenario sets
```
Comment (patrickhart) [inline bark/benchmark/tests/py_benchmark_runner_tests.py:59]: is the docstring correct?

File: bark/commons/BUILD [hunk 17]
```diff
@@ -3,11 +3,21 @@ package(default_visibility = ["//visibility:public"])
 cc_library(
     name = "commons",
     hdrs = [
-        "base_type.hpp"
+        "base_type.hpp",
```
Comment (patrickhart) [inline bark/commons/BUILD:6]: why include base_type here and not use it as dependency?

File: bark/commons/distribution/distribution.hpp [hunk 18]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
```
Comment (patrickhart) [inline bark/commons/distribution/distribution.hpp:1]: change to standard header

File: bark/commons/distribution/distributions_1d.hpp [hunk 19]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
```
Comment (patrickhart) [inline bark/commons/distribution/distributions_1d.hpp:1]: change to standard header

File: bark/commons/distribution/distributions_1d.hpp [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+#include "bark/commons/distribution/distribution.hpp"
+#include "bark/commons/params/params.hpp"
+
+
+namespace bark {
+namespace commons {
+
+class FixedValue : public Distribution {
+  public:
```
Comment (patrickhart) [inline bark/commons/distribution/distributions_1d.hpp:22]: format

File: bark/commons/distribution/distributions_1d.hpp [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+  auto probability = uniform_generator_(generator_);
+  auto sample = boost::math::quantile(dist_, probability);
+  return RandomVariate(1, sample);
+}
+
+using boost_normal = boost::math::normal_distribution<RandomVariableSupport>;
+using boost_uniform = boost::math::uniform_distribution<RandomVariableSupport>;
+
+template<>
+inline boost_uniform BoostDistribution1D<boost_uniform>::DistFromParams(const ParamsPtr& params) const {
```
Comment (patrickhart) [inline bark/commons/distribution/distributions_1d.hpp:86]: should avoid lines longer than 80 chars. makes it harder to review e.g. in github and not possible to view multiple codes side by side

File: bark/commons/distribution/multivariate_normal.hpp [hunk 22]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
+// Hart, Tobias Kessler
+//
+// This work is licensed under the terms of the MIT license.
+// For a copy, see <https://opensource.org/licenses/MIT>.
+
+#ifndef MODULES_COMMONS_DISTRIBUTION_MULTIVARIATE_NORMAL_HPP_
```
Comment (patrickhart) [inline bark/commons/distribution/multivariate_normal.hpp:7]: change all header guards

File: bark/commons/distribution/multivariate_normal.hpp [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+#include "bark/commons/params/params.hpp"
+
+
+namespace bark {
+namespace commons {
+
+// Based on https://stackoverflow.com/questions/6142576/sample-from-multivariate-normal-gaussian-distribution-in-c
+class MultivariateDistribution : public Distribution  {
+  public:
+    MultivariateDistribution(const ParamsPtr& params) : Distribution(params),
```
Comment (patrickhart) [inline bark/commons/distribution/multivariate_normal.hpp:23]: 2 spaces indent

File: bark/commons/params/BUILD [hunk 24]
```diff
@@ -8,11 +8,13 @@ cc_library(
...
     ],
     srcs = [
-        "setter_params.cpp"
+        "setter_params.cpp",
+        "params.cpp"
     ],
     visibility = ["//visibility:public"],
     deps = [
         "//bark/commons/util:util",
+        "//bark/commons/distribution:distribution",
```
Comment (patrickhart) [inline bark/commons/params/BUILD:17]: why do the params depend on the distribution?

File: bark/commons/tests/distribution_tests.cc [hunk 25]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick Hart, Tobias Kessler
```
Comment (patrickhart) [inline bark/commons/tests/distribution_tests.cc:1]: change

File: bark/commons/tests/distribution_tests.cc [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+  for (size_t i = 0; i < samples; ++i) {
+    auto sample = dist_normal.Sample()[0];
+    std_dev += abs(mean - sample) * abs(mean - sample);
+  }
+  EXPECT_NEAR(sqrt(std_dev/samples), 2.0f, 0.01);
+
+  
+}
+
+// TODO(fortiss): fill our this test
```
Comment (patrickhart) [inline bark/commons/tests/distribution_tests.cc:47]: remove

File: bark/commons/tests/distribution_tests.cc [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+#include <boost/shared_ptr.hpp>
+#include <boost/make_shared.hpp>
+#include "gtest/gtest.h"
+
+#include "bark/commons/params/setter_params.hpp"
+#include "bark/commons/distribution/distributions_1d.hpp"
+#include "bark/commons/distribution/multivariate_normal.hpp"
+
+
+// TODO(fortiss): fill our this test
```
Comment (patrickhart) [inline bark/commons/tests/distribution_tests.cc:19]: remove

File: bark/commons/tests/params_tests.cc [hunk 28]
```diff
@@ -1,5 +1,4 @@
-// Copyright (c) 2020 Julian Bernhard, Klemens Esterle, Patrick Hart and
-// Tobias Kessler
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick Hart, Tobias Kessler
```
Comment (patrickhart) [inline bark/commons/tests/params_tests.cc:1]: change

File: bark/models/behavior/action_store/behavior_action_store.cpp [hunk 29]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
```
Comment (patrickhart) [inline bark/models/behavior/action_store/behavior_action_store.cpp:1]: change

File: bark/models/behavior/action_store/behavior_action_store.cpp [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+namespace bark {
+namespace models {
+namespace behavior {
+
+using dynamic::Trajectory;
+
+
+ActionHash BehaviorActionStore::Store(const Action& action, const Trajectory& trajectory,
+                                      const BehaviorStatus& status) {
+  const auto action_hash = ActionToHash(action);
```
Comment (patrickhart) [inline bark/models/behavior/action_store/behavior_action_store.cpp:19]: why do we not just use a boost variant and use a hash here?
Without further documentation I do not understand why an action is stored like this

File: bark/models/behavior/action_store/behavior_action_store.hpp [hunk 31]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
```
Comment (patrickhart) [inline bark/models/behavior/action_store/behavior_action_store.hpp:1]: change

File: bark/models/behavior/action_store/behavior_action_store.hpp [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+
+
+using dynamic::Trajectory;
+using dynamic::Input;
+
+
+class BehaviorActionStore : public BehaviorModel {
+ public:
+  BehaviorActionStore(const commons::ParamsPtr& params)
+      : BehaviorModel(params),
```
Comment (patrickhart) [inline bark/models/behavior/action_store/behavior_action_store.hpp:30]: the clang formatter does not seem to work correctly. maybe talk to klemens about this?

File: bark/models/behavior/action_store/behavior_action_store.hpp [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+  std::tuple<Trajectory, Action, BehaviorStatus> Retrieve(const ActionHash& action_hash) const;
+
+  void MakeBehaviorActive(const ActionHash& action_hash) {
+    active_behavior_ = action_hash;
+  }
+
+  virtual Trajectory Plan(float delta_time, const bark::world::ObservedWorld& observed_world);
+
+  private:
+    std::unordered_map<ActionHash, std::tuple<Trajectory, Action, BehaviorStatus>> trajectory_store_;
```
Comment (patrickhart) [inline bark/models/behavior/action_store/behavior_action_store.hpp:46]: that is not really a neat data structure. consider using a struct

File: bark/models/behavior/behavior_model.hpp [hunk 34]
```diff
@@ -29,16 +29,52 @@ using dynamic::Trajectory;
...
+typedef boost::variant<DiscreteAction, Continuous1DAction, Input, LonLatAction> Action;
+typedef std::size_t ActionHash;
 
 typedef std::pair<State, Action> StateActionPair;
 typedef std::vector<StateActionPair> StateActionHistory;
 
 typedef std::vector<State> StateHistory;
 typedef std::vector<Action> ActionHistory;
 
+struct action_tostring_visitor : boost::static_visitor<std::string>
```
Comment (patrickhart) [inline bark/models/behavior/behavior_model.hpp:51]: formatting here as well

File: bark/models/behavior/idm/base_idm.hpp [hunk 35]
```diff
@@ -24,69 +24,77 @@ using bark::world::map::LaneCorridor;
...
+class BaseIDM : virtual public BehaviorModel {
+public:
+  explicit BaseIDM(const commons::ParamsPtr &params);
 
   virtual ~BaseIDM() {}
 
-  virtual Trajectory Plan(
-    float delta_time, const ObservedWorld& observed_world);
+  virtual Trajectory Plan(float delta_time,
+                          const ObservedWorld &observed_world);
```
Comment (patrickhart) [inline bark/models/behavior/idm/base_idm.hpp:44]: we agreed on `const ObservedWorld& observed_world`. please fix your formatter

Comment (patrickhart) [reply bark/models/behavior/idm/base_idm.hpp:44]: @cirrostratus1 had the same issue. maybe he can help you out with it.

File: bark/models/behavior/idm/idm_lane_tracking.cpp [hunk 36]
```diff
@@ -64,6 +65,10 @@ std::tuple<Trajectory, Action> BehaviorIDMLaneTracking::GenerateTrajectory(
       std::tie(acc, rel_distance) =
         GetTotalAcc(observed_world, rel_values, rel_distance, dt);
       BARK_EXPECT_TRUE(!std::isnan(acc));
+      // Set initial acceleration to maintain action value
+      if(i == 1) {
```
Comment (patrickhart) [inline bark/models/behavior/idm/idm_lane_tracking.cpp:69]: can you further elaborate this?

File: bark/models/behavior/idm/stochastic/idm_stochastic.cpp [hunk 37]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick
```
Comment (patrickhart) [inline bark/models/behavior/idm/stochastic/idm_stochastic.cpp:1]: change

File: bark/models/behavior/idm/stochastic/idm_stochastic.cpp [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "bark/models/behavior/idm/stochastic/idm_stochastic.hpp"
+
+
+namespace bark {
+namespace models {
+namespace behavior {
+
+
+BehaviorIDMStochastic::BehaviorIDMStochastic(const commons::ParamsPtr& params) : BehaviorIDMClassic(params),
```
Comment (patrickhart) [inline bark/models/behavior/idm/stochastic/idm_stochastic.cpp:15]: format

File: bark/models/behavior/idm/stochastic/idm_stochastic.hpp [hunk 39]
```diff
@@ -0,0 +1,48 @@
...
+  virtual ~BehaviorIDMStochastic() {}
+
+  virtual Trajectory Plan(float delta_time, const ObservedWorld& observed_world);
+
+  virtual std::shared_ptr<BehaviorModel> Clone() const;
+
+  void SampleParameters(); 
+
+  protected:
+    bark::commons::DistributionPtr param_dist_headway_;
```
Comment (patrickhart) [inline bark/models/behavior/idm/stochastic/idm_stochastic.hpp:30]: consider using bark::commons::DistributionPtr

File: bark/models/behavior/motion_primitives/macro_actions.cpp [hunk 40]
```diff
@@ -61,6 +61,7 @@ Trajectory BehaviorMPMacroActions::Plan(
 
   traj = selected_mp->Plan(delta_time, observed_world, target_corridor_);
 
+  this->SetLastAction(motion_primitives_.at(boost::get<DiscreteAction>(active_motion_))->GetLastAction());
```
Comment (patrickhart) [inline bark/models/behavior/motion_primitives/macro_actions.cpp:64]: format

File: bark/models/behavior/static_trajectory/behavior_static_trajectory.cpp [hunk 41]
```diff
@@ -8,24 +8,30 @@
...
 using bark::models::dynamic::StateDefinition;
 
-BehaviorStaticTrajectory::BehaviorStaticTrajectory(
-    const commons::ParamsPtr &params)
+BehaviorStaticTrajectory::BehaviorStaticTrajectory(const commons::ParamsPtr& params)
     : BehaviorModel(params, BehaviorStatus::NOT_STARTED_YET),
-      static_trajectory_(ReadInStaticTrajectory(params->GetListListFloat(
-          "static_trajectory",
-          "List of states that form a static trajectory to follow", {{}}))) {}
+      static_trajectory_(
```
Comment (patrickhart) [inline bark/models/behavior/static_trajectory/behavior_static_trajectory.cpp:21]: format

File: bark/models/tests/behavior_static_trajectory_test.cc [hunk 42]
```diff
@@ -68,6 +73,71 @@ TEST(behavior_static_trajectory_plan, plan) {
   EXPECT_EQ(expected, traj.row(2));
 }
 
+TEST(behavior_static_trajectory_plan, calculate_action) {
```
Comment (patrickhart) [inline bark/models/tests/behavior_static_trajectory_test.cc:76]: format

File: bark/models/tests/py_behavior_space_test.py [hunk 43]
```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2019 fortiss GmbH
```
Comment (patrickhart) [inline bark/models/tests/py_behavior_space_test.py:1]: change

File: bark/python_wrapper/commons/commons.cpp [hunk 44]
```diff
@@ -58,6 +65,23 @@ void python_commons(py::module m) {
 
     m.def("SetLogLevel", [](int level) { FLAGS_minloglevel = level; });
     m.def("SetVerboseLevel", [](int level) { FLAGS_v = level; });
+
+    py::class_<BaseType, std::shared_ptr<BaseType>>(m, "BaseType")
+      .def(py::init<const ParamsPtr&>())
+      .def_property_readonly("params", &BaseType::GetParams);
+
+    m.def("GLogInit", [](char* program_path, char* log_path,
```
Comment (patrickhart) [inline bark/python_wrapper/commons/commons.cpp:73]: can you document the novel logging concept?

File: bark/python_wrapper/models/behavior.cpp [hunk 45]
```diff
@@ -1,16 +1,17 @@
-// Copyright (c) 2020 Julian Bernhard, Klemens Esterle, Patrick Hart and
-// Tobias Kessler
+// Copyright (c) 2019 fortiss GmbH, Julian Bernhard, Klemens Esterle, Patrick Hart, Tobias Kessler
```
Comment (patrickhart) [inline bark/python_wrapper/models/behavior.cpp:1]: why to change the perfectly good header

File: bark/python_wrapper/models/behavior.cpp [hunk 46]
```diff
@@ -65,7 +67,7 @@ void python_behavior(py::module m) {
 
   py::class_<BehaviorIDMClassic,
              BehaviorModel,
-             shared_ptr<BehaviorIDMClassic>>(m, "BehaviorIDMClassic")
+             shared_ptr<BehaviorIDMClassic>>(m, "BehaviorIDMClassic", py::multiple_inheritance())
```
Comment (patrickhart) [inline bark/python_wrapper/models/behavior.cpp:70]: what does the muliple_inheritance do?

File: bark/python_wrapper/tests/py_glog_init.py [hunk 47]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class PickleTests(unittest.TestCase):
+
+    def test_init_glog_v1(self):
+        log_folder = os.path.abspath(os.path.join(os.getcwd(), "logs"))
+        if not os.path.exists(log_folder):
+          os.makedirs(log_folder)
+        print("Logging into: {}".format(log_folder))
+        bark.core.commons.GLogInit(sys.argv[0], log_folder, 1, False)
```
Comment (patrickhart) [inline bark/python_wrapper/tests/py_glog_init.py:15]: maybe document here also

File: bark/models/behavior/motion_primitives/primitives/primitive_const_acc_change_to_left.cpp [hunk 48]
```diff
@@ -7,7 +7,8 @@
 #include "primitive_const_acc_change_to_left.hpp"
 bark::models::behavior::primitives::PrimitiveConstAccChangeToLeft::
     PrimitiveConstAccChangeToLeft(const bark::commons::ParamsPtr& params)
-    : PrimitiveConstAccStayLane(params),
+    : BehaviorModel(params),
```
Comment (klemense1) [inline bark/models/behavior/motion_primitives/primitives/primitive_const_acc_change_to_left.cpp:10]: is this needed? PrimitiveConstAccStayLane will call this constructor anyway

File: bark/runtime/scenario/scenario_generation/config_readers/behavior_model_config_readers.py [hunk 49]
```diff
@@ -24,24 +25,75 @@ def create_from_config(self, config_param_object, road_corridor, agent_states,
...
+
+  def model_from_model_type(self, model_type, params):
+    bark_model = eval("{}(params)".format(model_type))
     return bark_model, params
 
   def get_param_servers(self):
     return self.param_servers
 
 
+class InteractionDataBehaviors(ConfigReaderBehaviorModels):
```
Comment (klemense1) [inline bark/runtime/scenario/scenario_generation/config_readers/behavior_model_config_readers.py:69]: do you have an example for this?

File: tools/debug/debug_settings.py [hunk 50]
```diff
@@ -0,0 +1,20 @@
+import sys
+import os
+
+
+
+cwd = os.getcwd()
```
Comment (klemense1) [inline tools/debug/debug_settings.py:6]: can you document this? does it needs to be run beforehand?

File: bark/models/behavior/motion_primitives/primitives/primitive_const_acc_change_to_left.cpp (hunk 48 above)
Comment (juloberno) [reply bark/models/behavior/motion_primitives/primitives/primitive_const_acc_change_to_left.cpp:10]: yes, to support virtual inheritance for the hypothesis interface

File: bark/models/behavior/idm/base_idm.hpp [hunk 51]
```diff
@@ -124,12 +133,15 @@ class BaseIDM : public BehaviorModel {
   bool brake_lane_end_;
   float brake_lane_end_enabled_distance_;
   float brake_lane_end_distance_offset_;
-};
 
+  // constant acceleration heuristic
+  // according chapter 11. Car-Following Models based on Driving Strategies
+  // in "Traffic Flow Dynamics" by M.Treiber and A.Kesting
+  float param_coolness_factor_;
```
[the remark this answers is not in the document]
Comment (juloberno) [reply bark/models/behavior/idm/base_idm.hpp:140]: yes, coolness factor 0.0f is the default

File: bark/models/behavior/behavior_model.hpp [hunk 52]
```diff
@@ -29,16 +29,52 @@ using dynamic::Trajectory;
...
+  Continuous1DAction acc_lon;
+
+  inline bool operator==(const LonLatAction& other) const {
+       return acc_lat == other.acc_lat && acc_lon == other.acc_lon;
+    }
+};
 using dynamic::Input;
 using models::dynamic::State;
-typedef boost::variant<DiscreteAction, Continuous1DAction, Input> Action;
+typedef boost::variant<DiscreteAction, Continuous1DAction, Input, LonLatAction> Action;
```
[the remark this answers is not in the document]
Comment (juloberno) [reply bark/models/behavior/behavior_model.hpp:42]: Why?

**Pull request merged**